dm chaos test meets ResourceFilesNotFoundError and fails

### What did you do?

Run chaos test, ref: https://github.com/pingcap/tiflow/actions/runs/3276912635/jobs/5393527322

### What did you expect to see?

_No response_

### What did you see instead?

- Test failed

`` Error: 0/18 21:21:03.721 +00:00] [ERROR] [main.go:87] ["run cases failed"] [error="different data for table `dmjob`.`tb0`"] [errorVerbose="different data for table `dmjob`.`tb0` ``

- DM dump and load worker don't finish, dump worker keeps failing

```go
[2022/10/18 21:16:14.538 +00:00] [INFO] [worker.go:280] ["worker finished exiting"] [job_id=8002fd57-a7df-45ce-b0f9-de595b1582df] [worker_id=64097aeb-382d-4649-b559-cce8fe3269bf] [caused="ResourceFilesNotFoundError: {"Ident":{"Executor":"chaos-executor-0-c0efd57c","WorkerID":"4059f82b-d657-4e42-99ad-8ca3f0e6a0ca","Name":"38303032666435372d613764662d343563652d623066392d6465353935623135383264662d7265706c6963612d3033"},"URI":"s3://engine-it/chaos-executor-0-c0efd57c/4059f82b-d657-4e42-99ad-8ca3f0e6a0ca/38303032666435372d613764662d343563652d623066392d6465353935623135383264662d7265706c6963612d3033","Details":""}"] [causedVerbose="ResourceFilesNotFoundError: {"Ident":{"Executor":"chaos-executor-0-c0efd57c","WorkerID":"4059f82b-d657-4e42-99ad-8ca3f0e6a0ca","Name":"38303032666435372d613764662d343563652d623066392d6465353935623135383264662d7265706c6963612d3033"},"URI":"s3://engine-it/chaos-executor-0-c0efd57c/4059f82b-d657-4e42-99ad-8ca3f0e6a0ca/38303032666435372d613764662d343563652d623066392d6465353935623135383264662d7265706c6963612d3033","Details":""}
github.com/pingcap/tiflow/engine/pkg/rpcerror.(*Prototype[...]).GenWithStack
        github.com/pingcap/tiflow/engine/pkg/rpcerror/normalize.go:101
github.com/pingcap/tiflow/engine/pkg/externalresource/internal/s3.(*FileManager).GetPersistedResource
        github.com/pingcap/tiflow/engine/pkg/externalresource/internal/s3/file_manager.go:125
github.com/pingcap/tiflow/engine/pkg/externalresource/broker.(*DefaultBroker).getPersistResource
        github.com/pingcap/tiflow/engine/pkg/externalresource/broker/broker.go:343
github.com/pingcap/tiflow/engine/pkg/externalresource/broker.(*DefaultBroker).OpenStorage
        github.com/pingcap/tiflow/engine/pkg/externalresource/broker/broker.go:174
github.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).OpenStorage
        github.com/pingcap/tiflow/engine/framework/worker.go:515
github.com/pingcap/tiflow/engine/executor/dm.(*dmWorker).setupStorage
        github.com/pingcap/tiflow/engine/executor/dm/worker.go:202
github.com/pingcap/tiflow/engine/executor/dm.(*dmWorker).InitImpl
        github.com/pingcap/tiflow/engine/executor/dm/worker.go:140
github.com/pingcap/tiflow/engine/framework.(*DefaultBaseWorker).Init
        github.com/pingcap/tiflow/engine/framework/worker.go:258
github.com/pingcap/tiflow/engine/framework/internal/eventloop.(*Runner[...]).doRun
        github.com/pingcap/tiflow/engine/framework/internal/eventloop/runner.go:84
github.com/pingcap/tiflow/engine/framework/internal/eventloop.(*Runner[...]).Run
        github.com/pingcap/tiflow/engine/framework/internal/eventloop/runner.go:59
github.com/pingcap/tiflow/engine/executor/worker.(*TaskRunner).launchTask.func1
        github.com/pingcap/tiflow/engine/executor/worker/task_runner.go:222
runtime.goexit
        runtime/asm_amd64.s:1594"] [duration=3.06168924s]
```



### Versions of the cluster

Dataflow Engine version (run `tiflow version`):

master@https://github.com/pingcap/tiflow/commit/9abde064c9a50aef77ffd3aa8be609d3265e4ed0

--- a/engine/pkg/externalresource/broker/broker.go
+++ b/engine/pkg/externalresource/broker/broker.go
@@ -344,6 +344,8 @@ func (b *DefaultBroker) getPersistResource(
 
 	if options.cleanBeforeOpen {
 		err := fm.RemoveResource(ctx, ident)
+		// LocalFileManager may return ErrResourceDoesNotExist, which can be 
+		// ignored because the resource no longer exists.
 		if err != nil && !derrors.ErrResourceDoesNotExist.Equal(err) {
 			return nil, err
 		}

--- a/engine/pkg/externalresource/broker/broker.go
+++ b/engine/pkg/externalresource/broker/broker.go
@@ -344,7 +344,7 @@ func (b *DefaultBroker) getPersistResource(
 
 	if options.cleanBeforeOpen {
 		err := fm.RemoveResource(ctx, ident)
-		// LocalFileManager may return ErrResourceDoesNotExist, which can be 
+		// LocalFileManager may return ErrResourceDoesNotExist, which can be
 		// ignored because the resource no longer exists.
 		if err != nil && !derrors.ErrResourceDoesNotExist.Equal(err) {
 			return nil, err

--- a/engine/pkg/externalresource/broker/broker.go
+++ b/engine/pkg/externalresource/broker/broker.go
@@ -344,7 +344,7 @@ func (b *DefaultBroker) getPersistResource(
 
 	if options.cleanBeforeOpen {
 		err := fm.RemoveResource(ctx, ident)
-		if err != nil {
+		if err != nil && !derrors.ErrResourceDoesNotExist.Equal(err) {
 			return nil, err
 		}
 		desc, err = fm.CreateResource(ctx, ident)

--- a/engine/pkg/externalresource/broker/broker.go
+++ b/engine/pkg/externalresource/broker/broker.go
@@ -340,10 +340,7 @@ func (b *DefaultBroker) getPersistResource(
 			WorkerID:    record.Worker,   /* creator id*/
 		},
 	}
-	desc, err := fm.GetPersistedResource(ctx, ident)
-	if err != nil {
-		return nil, err
-	}
+	var desc internal.ResourceDescriptor
 
 	if options.cleanBeforeOpen {
 		err := fm.RemoveResource(ctx, ident)
@@ -354,6 +351,12 @@ func (b *DefaultBroker) getPersistResource(
 		if err != nil {
 			return nil, err
 		}
+		return desc, nil
+	}
+
+	desc, err := fm.GetPersistedResource(ctx, ident)
+	if err != nil {
+		return nil, err
 	}
 	return desc, nil
 }

--- a/engine/pkg/externalresource/broker/broker_test.go
+++ b/engine/pkg/externalresource/broker/broker_test.go
@@ -21,6 +21,7 @@ import (
 
 	pb "github.com/pingcap/tiflow/engine/enginepb"
 	"github.com/pingcap/tiflow/engine/pkg/externalresource/internal/local"
+	"github.com/pingcap/tiflow/engine/pkg/externalresource/internal/s3"
 	"github.com/pingcap/tiflow/engine/pkg/externalresource/manager"
 	resModel "github.com/pingcap/tiflow/engine/pkg/externalresource/model"
 	"github.com/pingcap/tiflow/engine/pkg/tenant"
@@ -135,6 +136,43 @@ func TestBrokerOpenExistingStorage(t *testing.T) {
 	require.NoError(t, err)
 
 	local.AssertLocalFileExists(t, dir, "worker-2", resName, "1.txt")
+}
+
+func TestBrokerOpenExistingStorageWithOption(t *testing.T) {
+	t.Parallel()
+	fakeProjectInfo := tenant.NewProjectInfo("fakeTenant", "fakeProject")
+	brk, cli, _ := newBroker(t)
+	defer brk.Close()
+	mockS3FileManager, _ := s3.NewFileManagerForUT(t.TempDir(), brk.executorID)
+	brk.fileManagers[resModel.ResourceTypeS3] = mockS3FileManager
+
+	openStorageWithClean := func(resID resModel.ResourceID) {
+		// resource metadata exists
+		cli.On("QueryResource", mock.Anything,
+			&pb.QueryResourceRequest{ResourceKey: &pb.ResourceKey{JobId: "job-1", ResourceId: resID}}, mock.Anything).
+			Return(&pb.QueryResourceResponse{
+				CreatorExecutor: "executor-1",
+				JobId:           "job-1",
+				CreatorWorkerId: "worker-2",
+			}, nil)
+		hdl, err := brk.OpenStorage(
+			context.Background(),
+			fakeProjectInfo,
+			"worker-2",
+			"job-1",
+			resID, WithCleanBeforeOpen())
+		require.NoError(t, err)
+		require.Equal(t, resID, hdl.ID())
+		require.True(t, hdl.(*ResourceHandle).isPersisted.Load())
+	}
+
+	resIDs := []resModel.ResourceID{"/local/test-option", "/s3/test-option"}
+	for _, resID := range resIDs {
+		// resource does not exist, metadata exists
+		openStorageWithClean(resID)
+		// open again, resource exists, metadata exists
+		openStorageWithClean(resID)
+	}
 }
 
 func TestBrokerRemoveResource(t *testing.T) {

